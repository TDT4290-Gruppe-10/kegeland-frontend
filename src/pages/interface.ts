import { Chart } from 'chart.js';

export type SessionDataPoint = number | null;

export type Session = {
  id: string;
  userId: string;
  sensor: string;
  data: Record<string, SessionDataPoint[]>;
};

export type Sensor = {
  name: string;
  labels: string[];
};
