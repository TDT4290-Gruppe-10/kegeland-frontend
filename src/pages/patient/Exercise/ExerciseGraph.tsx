const ExerciseGraph = () => {
  return <h1>exerciseGraph</h1>;
};

export default ExerciseGraph;
