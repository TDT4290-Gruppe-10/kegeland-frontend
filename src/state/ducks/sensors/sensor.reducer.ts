import { createSlice } from "@reduxjs/toolkit";
import { getSessionData } from "./sensors.actions";

const initialState: any = {
  sessionData: {},
  loading: false,
  error: undefined,
};

const sensorSlice = createSlice({
  name: "sensors",
  initialState,
  reducers: {
    clearError: (state) => {
      state.error = undefined;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getSessionData.pending, (state) => {
        state.loading = true;
        state.error = undefined;
      })
      .addCase(getSessionData.fulfilled, (state, { payload }) => {
        state.loading = false;
        state.error = undefined;
        state.patients = payload;
      })
      .addCase(getSessionData.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      });
  },
});

export const { clearError } = sensorSlice.actions;

export const sensorReducer = sensorSlice.reducer;
