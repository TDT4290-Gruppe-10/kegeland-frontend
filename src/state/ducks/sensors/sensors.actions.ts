import { createAsyncThunk } from "@reduxjs/toolkit";

import { apiCaller } from "../../../utils/apiCaller";

export const getSessionData = createAsyncThunk(
  "sensors/session",
  async (sessionId: string) =>
    apiCaller<any>(`sensors/session/${sessionId}`, "GET").then(async (res) => {
      return res;
    })
);
