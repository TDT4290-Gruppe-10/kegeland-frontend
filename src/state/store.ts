import { combineReducers, configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import authReducer from './ducks/auth/auth.reducer';
import {patientReducer, singlePatientReducer} from './ducks/patients/patients.reducer';
import { sensorReducer } from './ducks/sensors/sensor.reducer';

const persistConfig = {
  key: 'root',
  version: 1,
  storage: storage,
  whitelist: ['auth'],
};

const rootReducer = combineReducers({
  auth: authReducer,
  patients: patientReducer,
  singlePatient: singlePatientReducer,
  sensorData: sensorReducer
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    })
      .concat(logger)
  },
  devTools: process.env.NODE_ENV !== 'production',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
