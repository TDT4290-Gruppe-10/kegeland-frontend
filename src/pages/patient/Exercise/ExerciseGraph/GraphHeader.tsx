import { HStack, IconButton, Tooltip } from '@chakra-ui/react';
import { capitalize } from 'lodash';
import moment from 'moment';
import {
  MdDock,
  MdDownload,
  MdOutlineWatchLater,
  MdSettings,
} from 'react-icons/md';

import LabeledValue from '../../../../components/LabeledValue';
import { Sensor } from '../../../../state/ducks/sensors/sensors.interface';
import { Session } from '../../../../state/ducks/sessions/sessions.interface';
import { getSessionDuration } from '../../../../utils/graph.utils';

type GraphHeaderProps = {
  sensor: Sensor;
  session: Session;
  toggleSettings: () => void;
};

const GraphHeader: React.FC<GraphHeaderProps> = ({
  sensor,
  session,
  toggleSettings,
}) => {
  return (
    <HStack justifyContent="space-between">
      <HStack spacing={12}>
        <LabeledValue
          label="Date of exercise"
          value={moment.unix(session.createdAt).format('lll')}
        />

        <LabeledValue
          label="Device"
          icon={MdDock}
          value={capitalize(sensor.name)}
        />

        <LabeledValue
          label="Duration"
          icon={MdOutlineWatchLater}
          value={`${getSessionDuration(session).toFixed(0)}s`}
        />
      </HStack>
      <HStack spacing={0}>
        <IconButton
          aria-label="Open graph settings"
          icon={<MdSettings />}
          variant="unstyled"
          fontSize={22}
          color="gray.700"
          _hover={{ color: 'gray.600' }}
          onClick={toggleSettings}
        />
        <Tooltip label="Download data as csv">
          <IconButton
            aria-label="Download exercise data"
            icon={<MdDownload />}
            variant="unstyled"
            fontSize={22}
            color="gray.700"
            _hover={{ color: 'gray.600' }}
            onClick={toggleSettings}
          />
        </Tooltip>
      </HStack>
    </HStack>
  );
};

export default GraphHeader;
