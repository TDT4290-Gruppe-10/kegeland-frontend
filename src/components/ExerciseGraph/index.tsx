import { Chart } from 'react-chartjs-2';
import zoomPlugin from 'chartjs-plugin-zoom';
import { Chart as ChartJS } from 'chart.js';
import 'chart.js/auto';
import { useDisclosure, Divider } from '@chakra-ui/react';
import 'chartjs-adapter-moment';

import useGraphProfile from '../../hooks/useGraphProfile';
import { Sensor } from '../../state/ducks/sensors/sensors.interface';
import { Session } from '../../state/ducks/sessions/sessions.interface';

import GraphOptionsModal from './GraphOptionsModal';
import GraphHeader from './GraphHeader';

ChartJS.register(zoomPlugin);

type ExerciseGraphProps = {
  sensor: Sensor;
  session: Session;
};

const ExerciseGraph: React.FC<ExerciseGraphProps> = ({ sensor, session }) => {
  const { options, chartData, chartRef, updatePlot, resetPlot, updateXAxis } =
    useGraphProfile(sensor, session);
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <>
      {chartData && (
        <>
          <GraphHeader
            sensor={sensor}
            session={session}
            toggleSettings={onOpen}
          />
          <Divider borderColor="gray.300" marginY={4} />
          <div style={{ width: '99%' }}>
            <Chart
              height={200}
              ref={chartRef}
              type="line"
              options={options}
              data={chartData}
            />
          </div>
        </>
      )}
      <GraphOptionsModal
        sensor={sensor}
        isOpen={isOpen}
        onClose={onClose}
        updatePlot={updatePlot}
        resetPlot={resetPlot}
        updateXAxis={updateXAxis}
      />
    </>
  );
};

export default ExerciseGraph;
