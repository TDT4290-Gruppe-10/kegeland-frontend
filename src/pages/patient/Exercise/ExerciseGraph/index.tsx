import { Chart } from 'react-chartjs-2';
import zoomPlugin from 'chartjs-plugin-zoom';
import { Chart as ChartJS } from 'chart.js';
import 'chart.js/auto';
import { Container, Spinner, useDisclosure, Divider } from '@chakra-ui/react';
import 'chartjs-adapter-moment';

import useGraphProfile from '../../../../hooks/useGraphProfile';
import { WithPatientExercise } from '../../../../hoc/withPatientExercise';

import GraphOptionsModal from './GraphOptionsModal';
import GraphHeader from './GraphHeader';

ChartJS.register(zoomPlugin);

type ExerciseGraphProps = WithPatientExercise;

const ExerciseGraph: React.FC<ExerciseGraphProps> = ({ sensor, session }) => {
  const { options, chartData, chartRef, updatePlot, resetPlot, updateXAxis } =
    useGraphProfile(sensor, session);
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <Container
      width="full"
      maxW="container.xl"
      minH="2xl"
      paddingY={4}
      backgroundColor="white"
      borderRadius={5}>
      {chartData ? (
        <>
          <GraphHeader
            sensor={sensor}
            session={session}
            toggleSettings={onOpen}
          />
          <Divider borderColor="gray.300" marginY={4} />
          <Chart
            ref={chartRef}
            type="line"
            options={options}
            data={chartData}
          />
        </>
      ) : (
        <Spinner
          justifySelf={'center'}
          justifyItems={'center'}
          textAlign="center"
          alignSelf={'center'}
        />
      )}
      <GraphOptionsModal
        sensor={sensor}
        isOpen={isOpen}
        onClose={onClose}
        updatePlot={updatePlot}
        resetPlot={resetPlot}
        updateXAxis={updateXAxis}
      />
    </Container>
  );
};

export default ExerciseGraph;
