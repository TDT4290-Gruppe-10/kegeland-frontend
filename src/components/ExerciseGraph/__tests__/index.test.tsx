import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import ExerciseGraph from '..';
import { SensorType } from '../../../state/ducks/sensors/sensors.interface';
import { store } from '../../../state/store';
import { session } from '../../mocks/sessions.mock';

describe('Test exrecise index', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ExerciseGraph
            sensor={{
              id: 'sensorId',
              name: SensorType.FEMFIT,
              labels: ['s1', 's2', 's3'],
            }}
            session={session}
          />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
