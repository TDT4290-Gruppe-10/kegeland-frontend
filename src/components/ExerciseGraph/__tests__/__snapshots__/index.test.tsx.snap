// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test exrecise index renders correctly 1`] = `null`;
